### ISSUE TITLE
option: Unclear best practice and documentation for option.ImpersonateCredentials

### ISSUE THREAD

**Issue by joshk0** [googleapis/google-api-go-client#731]
In #625, this comment was added to better explain the role of scopes in impersonation for the new option.ImpersonateCredentials feature.

```
// The resulting impersonated credential will either have the default scopes of
// the client being instantiating or the scopes from WithScopes if provided.
// Scopes are required for creating impersonated credentials, so if this option
// is used while not using a NewClient/NewService function, WithScopes must also
// be explicitly passed in as well.
```

Despite this comment, in my code, if I do this:

```go
 bq, err := bigquery.NewClient(ctx, c.accessProjectID, option.ImpersonateCredentials(email))
```

and then try to use `bq`, I always get the error of a 403 from the call to [generateAccessToken](https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateAccessToken) and insufficient scopes described as the cause at [this line](https://github.com/googleapis/google-api-go-client/blob/ba82a99e06960678387a0b6c6ef01a0ed14c202a/internal/impersonate/impersonate.go#L113).

For the same code to work, I have to do this:

```go
 bq, err := bigquery.NewClient(ctx, c.accessProjectID, option.WithScopes("https://www.googleapis.com/auth/cloud-platform"), option.ImpersonateCredentials(email))
```

Now I can use `bq` as the impersonated service account as I expect.

I am raising this issue because the documentation seems unclear with regards to the need to pass the scopes in even though I do happen to be inside a NewClient() call. Also, it seems concerning that this will force the resulting token to be too powerful, even though I am creating it just for BigQuery purposes.

It seems like the ideal behavior would be that a native token with iam scopes is requested to perform the impersonation, then the regular requested scopes for the client or service are requested for the impersonated token. This way, tokens will be minimally scoped.

In summary I am proposing one of these outcomes:

* It is properly documented that WithScopes() must be passed to get ImpersonateCredentials() to work intuitively.
* A token with the proper iam or cloud-platform scopes is requested only for the purpose of generating tokens; the client's default scopes are then requested for the impersonated token. In other words, the call to WithScopes() shouldn't affect impersonation.

I'm also happy to be hit with the cluebat in case I'm not understanding the documentation or interface correctly. Do at least consider it a vote that the usage as documented is unclear.

cc @salrashid123 as I only knew the hint about WithScopes by reading his helpful repo, salrashid123/oauth2 👍🏼

**Comment by codyoss** [comment]
Hey @joshk0,

Thank you for the detailed feedback, this was very helpful. 

>I always get the error of a 403 from the call to generateAccessToken and insufficient scopes described as the cause at this line.

Most of our clients, by default, set the `https://www.googleapis.com/auth/cloud-platform` scope. It appears bigquery does not though which is why you are having some trouble here. 

> Also, it seems concerning that this will force the resulting token to be too powerful, even though I am creating it just for BigQuery purposes.

This is a good point. Although the scope may be broad the token is still limited by what polices the service account being impersonated has been granted.

> It seems like the ideal behavior would be that a native token with iam scopes is requested to perform the impersonation, then the regular requested scopes for the client or service are requested for the impersonated token. This way, tokens will be minimally scoped.

I agree, this does sound like a good path forward. Let me do some investigating here and I will report back.

Thanks again!

**Comment by codyoss** [comment]
Please try out our impersonate preview and feel free to provide any feedback: https://github.com/googleapis/google-api-go-client/releases/tag/v0.44.0-impersonate-preview

**Comment by codyoss** [comment]
Now in a main release: https://github.com/googleapis/google-api-go-client/releases/tag/v0.46.0
